mmtrack/models/trackers/ocsort_tracker.py

```diff
@@ -0,0 +1,457 @@
+                Defaults to 0.5.
+
+        Returns:
+            tuple(int): The assigning ids.
+
+        OC-SORT uses velocity consistency besides IoU for association
+        """
+        # get track_bboxes
+        track_bboxes = np.zeros((0, 4))
+        for id in ids:
+            track_bboxes = np.concatenate(
+                (track_bboxes, self.tracks[id].mean[:4][None]), axis=0)
+        track_bboxes = torch.from_numpy(track_bboxes).to(det_bboxes)
+        track_bboxes = bbox_cxcyah_to_xyxy(track_bboxes)
+
+        # compute distance
+        ious = bbox_overlaps(track_bboxes, det_bboxes[:, :4])
+        if weight_iou_with_det_scores:
+            ious *= det_bboxes[:, 4][None]
+        dists = (1 - ious).cpu().numpy()
+
+        if len(ids) > 0 and len(det_bboxes) > 0:
+            track_velocities = torch.stack(
+                [self.tracks[id].velocity for id in ids]).to(det_bboxes.device)
+            k_step_observations = torch.stack([
+                self.k_step_observation(self.tracks[id]) for id in ids
+            ]).to(det_bboxes.device)
+            valid1 = track_velocities.sum(dim=1) != -2
+            valid2 = k_step_observations.sum(dim=1) != -4
+            valid = valid1 & valid2
+
+            vel_to_match = self.vel_direction_batch(k_step_observations[:, :4],
+                                                    det_bboxes[:, :4])
+            track_velocities = track_velocities[:, None, :].repeat(
+                1, det_bboxes.shape[0], 1)
+
+            angle_cos = (vel_to_match * track_velocities).sum(dim=-1)
+            angle_cos = torch.clamp(angle_cos, min=-1, max=1)
+            angle = torch.acos(angle_cos)  # [0, pi]
+            norm_angle = (angle - np.pi / 2.) / np.pi  # [-0.5, 0.5]
+            valid_matrix = valid[:, None].int().repeat(1, det_bboxes.shape[0])
+            # set non-valid entries 0
+            valid_norm_angle = norm_angle * valid_matrix
+
+            dists += valid_norm_angle.cpu().numpy() * self.vel_consist_weight
+
+        # bipartite match
+        if dists.size > 0:
+            cost, row, col = lap.lapjv(
+                dists, extend_cost=True, cost_limit=1 - match_iou_thr)
+        else:
+            row = np.zeros(len(ids)).astype(np.int32) - 1
+            col = np.zeros(len(det_bboxes)).astype(np.int32) - 1
+        return row, col
+
+    def last_obs(self, track):
+        """extract the last associated observation."""
+        for bbox in track.obs[::-1]:
+            if bbox is not None:
+                return bbox
+
+    def ocr_assign_ids(self,
+                       track_obs,
+                       det_bboxes,
+                       weight_iou_with_det_scores=False,
+                       match_iou_thr=0.5):
+        """association for Observation-Centric Recovery.
+
+        As try to recover tracks from being lost whose estimated velocity is
+        out- to-date, we use IoU-only matching strategy.
```

Comment:
missing docstring.

We need full docstring if a function has more than 2 args or more than10 lines.